src/main/java/com/microsoft/jenkins/azuread/AzureSecurityRealm.java

```diff
@@ -296,10 +349,12 @@ private String getRootUrl() {
     }
 
     @DataBoundConstructor
-    public AzureSecurityRealm(String tenant, String clientId, Secret clientSecret, int cacheDuration) {
+    public AzureSecurityRealm(String tenant, String clientId, Secret clientSecret, Secret clientCertificate, String credentialType, int cacheDuration) {
```

Comment:
The DataBoundConstructor is part of the public API, in order to add new fields they should be added as a `DataBoundSetter` from a setter method. 

That way people using configuration as code won't all need to change their configuration.

Reply:
Could you please elaborate? 
It means we should delete the new parameters from "DataBoundConstructor" to avoid breaking changes to JCasC users? 

Set these parameters under "@DataBoundSetter" will be enough?

Reply:
yes exactly

Reply:
And that way JCasC users will still be able to use the new parameters in their config? (e.x using certificate instead of secret using JCasC)

Reply:
yes you can add another test to verify that

Reply:
@timja hi I've deleted the parameters from the 'DataBoundConstructor' , added DataBoundSetter for the new methods and adjusted the test accordingly please review.

One thing, for the JCasC support, I'm trying to validate using the ConfigAsCodeTest I think I am missing something locally as I cannot run the tests, I'm receiving the following error when I try to run them:

 `"=== Starting com.microsoft.jenkins.azuread.integrations.casc.ConfigAsCodeTest
   0.026 [id=25]        INFO    o.jvnet.hudson.test.WarExploder#explode: Picking up existing exploded jenkins.war at /Users/itzikbekel/JenkinsRepos/azure-ad-plugin/target/jenkins-for-test
   0.172 [id=25]        INFO    o.jvnet.hudson.test.JenkinsRule#createWebServer: Running on http://localhost:51037/jenkins/
   0.331 [id=41]        INFO    jenkins.InitReactorRunner$1#onAttained: Started initialization
   0.337 [id=65]        INFO    jenkins.InitReactorRunner$1#onAttained: Listed all plugins
   1.051 [id=41]        INFO    jenkins.InitReactorRunner$1#onAttained: Prepared all plugins
   1.056 [id=57]        INFO    jenkins.InitReactorRunner$1#onAttained: Started all plugins
   1.058 [id=39]        INFO    jenkins.InitReactorRunner$1#onAttained: Augmented all extensions
   1.445 [id=65]        WARNING h.ExtensionFinder$GuiceFinder$FaultTolerantScope$1#error: Failed to instantiate Key[type=org.jenkinsci.main.modules.instance_identity.PageDecoratorImpl, annotation=[none]]; skipping this component
java.lang.IllegalStateException: Couldn't locate either of bouncy castle FIPS or non fips provider, available providers areSUN,SunRsaSign,SunEC,SunJSSE,SunJCE,SunJGSS,SunSASL,XMLDSig,SunPCSC,JdkLDAP,JdkSASL,Apple,SunPKCS11.
        at jenkins.bouncycastle.api.PEMEncodable.<clinit>(PEMEncodable.java:96)
Caused: java.lang.ExceptionInInitializerError
        at org.jenkinsci.main.modules.instance_identity.pem.PEMHelper.encodePEM(PEMHelper.java:72)
        at org.jenkinsci.main.modules.instance_identity.InstanceIdentity.write(InstanceIdentity.java:103)
        at org.jenkinsci.main.modules.instance_identity.InstanceIdentity.<init>(InstanceIdentity.java:69)
        at org.jenkinsci.main.modules.instance_identity.InstanceIdentity.<init>(InstanceIdentity.java:43)
        at org.jenkinsci.main.modules.instance_identity.PageDecoratorImpl.<init>(PageDecoratorImpl.java:18)
        at org.jenkinsci.main.modules.instance_identity.PageDecoratorImpl$$FastClassByGuice$$acfecfa.GUICE$TRAMPOLINE(<generated>)
        at org.jenkinsci.main.modules.instance_identity.PageDecoratorImpl$$FastClassByGuice$$acfecfa.apply(<generated>)
        at com.google.inject.internal.DefaultConstructionProxyFactory$FastClassProxy.newInstance(DefaultConstructionProxyFactory.java:82)
        at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:114)
        at com.google.inject.internal.ConstructorInjector.access$000(ConstructorInjector.java:33)
        at com.google.inject.internal.ConstructorInjector$1.call(ConstructorInjector.java:98)
        at com.google.inject.internal.ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:109)
        at hudson.ExtensionFinder$GuiceFinder$SezpozModule.onProvision(ExtensionFinder.java:610)
        at com.google.inject.internal.ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:117)
        at com.google.inject.internal.ProvisionListenerStackCallback.provision(ProvisionListenerStackCallback.java:66)
        at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:93)
        at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:300)
        at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
Caused: com.google.inject.ProvisionException: Unable to provision, see the following errors:

1) [Guice/ErrorInjectingConstructor]: ExceptionInInitializerError
  at PageDecoratorImpl.<init>(PageDecoratorImpl.java:17)

Learn more:
  https://github.com/google/guice/wiki/ERROR_INJECTING_CONSTRUCTOR

1 error

======================
Full classname legend:
======================
PageDecoratorImpl:           "org.jenkinsci.main.modules.instance_identity.PageDecoratorImpl"
========================
End of classname legend:
========================

        at com.google.inject.internal.InternalProvisionException.toProvisionException(InternalProvisionException.java:251)
        at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:43)
        at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:169)
        at hudson.ExtensionFinder$GuiceFinder$FaultTolerantScope$1.get(ExtensionFinder.java:445)
        at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
        at com.google.inject.internal.InjectorImpl$1.get(InjectorImpl.java:1148)
        at hudson.ExtensionFinder$GuiceFinder._find(ExtensionFinder.java:403)
        at hudson.ExtensionFinder$GuiceFinder.find(ExtensionFinder.java:394)
        at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:344)
        at hudson.ExtensionList.load(ExtensionList.java:384)
        at hudson.ExtensionList.ensureLoaded(ExtensionList.java:320)
        at hudson.ExtensionList.iterator(ExtensionList.java:172)
        at hudson.diagnosis.NullIdDescriptorMonitor.verify(NullIdDescriptorMonitor.java:72)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:109)
        at hudson.init.TaskMethodFinder$TaskImpl.run(TaskMethodFinder.java:185)
        at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:305)
        at jenkins.model.Jenkins$5.runTask(Jenkins.java:1166)
        at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:221)
        at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:120)
        at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:68)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)
   1.669 [id=53]        INFO    jenkins.InitReactorRunner$1#onAttained: System config loaded
   1.724 [id=46]        INFO    jenkins.InitReactorRunner$1#onAttained: System config adapted
   1.724 [id=61]        INFO    jenkins.InitReactorRunner$1#onAttained: Loaded all jobs
   1.728 [id=61]        INFO    jenkins.InitReactorRunner$1#onAttained: Configuration for all jobs updated
   1.740 [id=54]        INFO    jenkins.InitReactorRunner$1#onAttained: Completed initialization

"`

Do you know how it can be resolved?

Reply:
How are you running them?

Generally if you're having troubles in your IDE then you can run it from the command line,

`mvn test -Dtest=...`

Reply:
Hi @timja , tried to run different ways, unfortunately I cannot get the test working and therefore I'm not able to debug.
Could you please have a look and see what is missing in the test I've added? 
Seems like the certificate is not asserted using the JCasC config
see [link](https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fazure-ad-plugin/detail/PR-596/9/pipeline/)

Reply:
> Hi @timja , tried to run different ways, unfortunately I cannot get the test working and therefore I'm not able to debug. Could you please have a look and see what is missing in the test I've added? Seems like the certificate is not asserted using the JCasC config see [link](https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fazure-ad-plugin/detail/PR-596/9/pipeline/)

@timja NVM, I managed to solve it by splitting the tests to separate files.
Please review